python/iceberg/api/expressions/expression_parser.py

```diff
@@ -0,0 +1,157 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+
+from pyparsing import (
+    alphanums,
+    alphas,
+    CaselessKeyword,
+    delimitedList,
+    Group,
+    infixNotation,
+    oneOf,
+    opAssoc,
+    pyparsing_common as ppc,
+    quotedString,
+    Word
+)
+
+
+_logger = logging.getLogger(__name__)
+
+AND, OR, IN, IS, NOT, NULL, BETWEEN = map(
+    CaselessKeyword, "and or in is not null between".split()
+)
+NOT_NULL = NOT + NULL
+
+ident = Word(alphas, alphanums + "_$").setName("identifier")
+columnName = delimitedList(ident, ".", combine=True).setName("column name")
+
+binop = oneOf("= == != < > >= <= eq ne lt le gt ge <>", caseless=False)
+realNum = ppc.real()
+intNum = ppc.signed_integer()
+
+columnRval = (realNum
+              | intNum
+              | quotedString
+              | columnName)  # need to add support for alg expressions
+whereCondition = Group(
+    (columnName + binop + columnRval)
+    | (columnName + IN + Group("(" + delimitedList(columnRval) + ")"))
+    | (columnName + IS + (NULL | NOT_NULL))
+    | (columnName + BETWEEN + columnRval + AND + columnRval)
+
+)
+
+whereExpression = infixNotation(
+    Group(whereCondition
+          | NOT + whereCondition
+          | NOT + Group('(' + whereCondition + ')')
+          | NOT + columnName),
+    [(NOT, 1, opAssoc.LEFT), (AND, 2, opAssoc.LEFT), (OR, 2, opAssoc.LEFT), (IS, 2, opAssoc.LEFT)],
+)
+
+op_map = {"=": "eq",
+          "==": "eq",
+          "eq": "eq",
+          ">": "gt",
+          "gt": "gt",
+          ">=": "gte",
+          "gte": "gte",
+          "<": "lt",
+          "lt": "lt",
+          "<=": "lte",
+          "lte": "lte",
+          "!": "not",
+          "not": "not",
+          "!=": "neq",
+          "<>": "neq",
+          "neq": "neq",
+          "||": "or",
+          "or": "or",
+          "&&": "and",
+          "and": "and",
+          "in": "in",
+          "between": "between",
+          "is": "is"}
```

Comment:
Is this derived from another parser or example? If so, we should make sure we note it so we can update the license documentation.

Reply:
Updated with some sourcing for the basic parsing grammar